### PR TITLE
#1016: Fix and tests for a type inference error

### LINKED ISSUE #1016: Incorrectly inferred type with :-
If `Gimmie()` is a function that returns an `Option<int>`, then the expression

``` dafny
var index :- GetIndex(log); Some(100.0)
```

should have type `Option<real>`. However, type inference sometimes computes a different type. For example, if `s` is of some type `State`, then type inference comes up with `Option<State>` as the type of both LHS and RHS of the following equality:

``` dafny
Some(s) == var index :- GetIndex(log); Some(100.0)
```

which is bogus.

In the repro below, the equality above is in method `Test5`. The other tests show variations.

``` dafny
datatype Option<T> = Some(value: T) | None {
  predicate method IsFailure() {
    None?
  }
  function method PropagateFailure<U>(): Option<U>
    requires None?
  {
    None
  }
  function method Extract(): T
    requires Some?
  {
    value
  }
}

type Process(==)

type State(==)

datatype Event = Event(process: Process, state: State)

function method Find(process: Process, log: seq<Event>): (r: Option<nat>)
  ensures r.Some? ==> r.value < |log|

function method Gimmie(): Option<int>

method Test0(process: Process, m: map<Process, State>, log: seq<Event>)
  requires process in m.Keys
{
  // x has type Option<State>
  var x :=
    Some(m[process]);
  // y has type Option<Process>
  var y :=
    var L := Find(process, log);
    if L == None then None else Some(log[L.value].process);
  // z's RHS is a different way of writing y's RHS
  // so, z also has type Option<Process>
  var z :=
    var last :- Find(process, log);
    Some(log[last].process);

  // Without the type errors in the other methods, the following assertion
  // type checks and verifies.
  assert y == z;
}

method Test1(process: Process, m: map<Process, State>, log: seq<Event>)
  requires process in m.Keys
{
  // The definitions of x, y, and z are exactly like in Test0
  var x :=
    Some(m[process]);
  var y :=
    var L := Find(process, log);
    if L == None then None else Some(log[L.value].process);
  var z :=
    var last :- Find(process, log);
    Some(log[last].process);

  // The following line gives the expected type error
  var b := x == y;  // error: incompatible types Option<State> and Option<Process>
}

method Test2(process: Process, m: map<Process, State>, log: seq<Event>)
  requires process in m.Keys
{
  // The definitions of x, y, and z are exactly like in Test0
  var x :=
    Some(m[process]);
  var y :=
    var L := Find(process, log);
    if L == None then None else Some(log[L.value].process);
  var z :=
    var last :- Find(process, log);
    Some(log[last].process);

  var c := x == z;  // BOGUS: this should give a type error, but it does not
  // In fact, using /rprint shows that the type of z is inferred to be Option<State>
}

method Test3(process: Process, m: map<Process, State>, log: seq<Event>)
  requires process in m.Keys
{
  // The definitions of x, y, and z are like in Test0, except the types are given
  // explicitly here.
  var x: Option<State> :=
    Some(m[process]);
  var y: Option<Process> :=
    var L := Find(process, log);
    if L == None then None else Some(log[L.value].process);
  var z: Option<Process> :=
    var last :- Find(process, log);
    Some(log[last].process);

  // With the explicit types above, the assignments above type check, and
  // the following line gives the expected type error.
  var c := x == z;  // error: incompatible types Option<State> and Option<Process>
}

method Test4(process: Process, m: map<Process, State>, log: seq<Event>)
  requires process in m.Keys
{
  // The definitions of x, y, and z are exactly like in Test0
  var x :=
    Some(m[process]);
  var y :=
    var L := Find(process, log);
    if L == None then None else Some(log[L.value].process);
  var z :=
    var last :- Find(process, log);
    Some(100);

  var c := x == z;  // BOGUS: this should give a type error, but it does not
  // In fact, using /rprint shows that the type of z is inferred to be Option<State>
}

method Test5(s: State)
{
  // The definitions of x, y, and z are exactly like in Test0
  var x :=
    Some(s);
  var z :=
    var n :- Gimmie();
    Some(100.0);

  var c := x == z;  // BOGUS: this should give a type error, but it does not
  // In fact, using /rprint shows that the type of z is inferred to be Option<State>
}
```

### PR DESCRIPTION
**Pull request by davidcok**
Fixes #1016

### CODE REVIEW

File: Test/git-issues/git-issue-1016.dfy [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    Some(m[process]);
+  var y :=
+    var L := Find(process, log);
+    if L == None then None else Some(log[L.value].process);
+  var z :=
+    var last :- Find(process, log);
+    Some(100);
+
+  var c := x == z;  // ERROR: this should give a type error
+  // In fact, using /rprint shows that the type of z is inferred to be Option<State>
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-1016.dfy:119]:
```suggestion
```

File: Test/git-issues/git-issue-1016.dfy [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+{
+  // The definitions of x, y, and z are exactly like in Test0
+  var x :=
+    Some(s);
+  var z :=
+    var n :- Gimmie();
+    Some(100.0);
+
+  var c := x == z;  // ERROR: this should give a type error
+  // In fact, using /rprint shows that the type of z is inferred to be Option<State>
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-1016.dfy:132]:
```suggestion
```

File: Test/git-issues/git-issue-1016a.dfy [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+method Test5(s: State)
+{
+  // The definitions of x, y, and z are exactly like in Test0
+  var x := Some(s);
+  var z :=
+    var n :- Gimmie();
+    Some(100.0);
+
+  var c := x == z;  // BOGUS: this should give a type error, but it does not
+  // In fact, using /rprint shows that the type of z is inferred to be Option<State>
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-1016a.dfy:35]:
```suggestion
  var c := x == z;  // BOGUS: this should give a type error
```

**Review by RustanLeino — CHANGES_REQUESTED** [review]

File: Test/git-issues/git-issue-1016.dfy (hunk 1 above)
Comment (davidcok) [reply Test/git-issues/git-issue-1016.dfy]: Fixed

File: Test/git-issues/git-issue-1016a.dfy (hunk 3 above)
Comment (davidcok) [reply Test/git-issues/git-issue-1016a.dfy]: Fixed

File: Test/git-issues/git-issue-1016.dfy (hunk 2 above)
Comment (davidcok) [reply Test/git-issues/git-issue-1016.dfy]: Fixed

**Review by RustanLeino — APPROVED** [review]

**Pull request merged**